README.md
# smtp
[![Pub](https://img.shields.io/pub/v/smtp.svg)](https://pub.dartlang.org/packages/smtp)
[![build status](https://travis-ci.org/thosakwe/smtp.svg)](https://travis-ci.org/thosakwe/smtp)

An
[SMTP 5321](https://tools.ietf.org/html/rfc5321)
server implementation in Dart.

## Usage
This SMTP server should not be used as your front-facing server,
especially in production. Consider setting up Postfix to relay e-mail
to this server; that way, you can be sure that only well-formed e-mail
messages will touch your Dart code.

Note that this includes no spam protection, etc., and is simply
a plain SMTP server library.

```dart

main() async {
  var server = await SmtpServer.bind('127.0.0.1', 0);

  await for (var mailObject in server.mailObjects) {
    print(mailObject.envelope.originatorAddress);
    print(mailObject.envelope.headers.cc);
    mailObject.close(statusCode: 221, reasonPhrase: 'Bye!');
  }
}
```

CHANGELOG.md
# 1.0.0+1
* Made the description longer for Pub (not sure why this knocked off an entire 20% from the score...).

# 1.0.0
* First release.

test/all_test.dart
import 'dart:convert' show LineSplitter;
import 'dart:io';
import 'package:async/async.dart';
import 'package:dart2_constant/convert.dart';
import 'package:mailer/mailer.dart';
import 'package:smtp/smtp.dart';
import 'package:test/test.dart';

void main() {
  SmtpServer server;

  setUp(() async {
    server = await SmtpServer.bind('127.0.0.1', 0);
  });

  tearDown(() async {
    await server.close();
  });

  test('plain text', () async {
    var smtp = new SmtpTransport(new SmtpOptions()
      ..hostName = server.address.address
      ..port = server.port);

    var envelope = new Envelope()
      ..from = 'foo@bar.com'
      ..recipients.add('someone@somewhere.com')
      ..recipients.add('someone_else@somewhere.com')
      ..bccRecipients.add('hidden@recipient.com')
      ..ccRecipients.add('also_hidden@recipient.com')
      ..ccRecipients.add('also_also_hidden@recipient.com')
      ..subject = 'Testing the Dart Mailer library 語'
      ..text = 'This is a cool email message. Whats up? 語';
    //..html = '<h1>Test</h1><p>Hey!</p>';

    smtp.send(envelope);

    var mailObject = await server.mailObjects.first;
    await mailObject.close();

    expect(mailObject.supportsSmtpExtensions, isTrue);
    expect(mailObject.envelope.originatorAddress, envelope.from);
    expect(mailObject.envelope.recipientAddresses, hasLength(5));
    print(mailObject.envelope.headers.toMap());
    // expect(mailObject.envelope.headers.bcc, envelope.bccRecipients);
    expect(mailObject.envelope.headers.cc, envelope.ccRecipients);
    expect(mailObject.envelope.headers.contentType.mimeType, 'multipart/mixed');
    expect(mailObject.envelope.headers.subject, envelope.subject);
  });
  test('read multipart', () async {
    var smtp = new SmtpTransport(new SmtpOptions()
      ..hostName = server.address.address
      ..port = server.port);

    var envelope = new Envelope()
      ..from = 'foo@bar.com'
      ..recipients.add('someone@somewhere.com')
      ..subject = 'Huh HTML'
      ..html = '<h1>Test</h1><p>Hey!</p>';

    smtp.send(envelope);

    var mailObject = await server.mailObjects.first;
    await mailObject.close();

    print(mailObject.content);

    var part = await mailObject.mimeMultiparts.first;
    expect(part.headers['Content-Type'], 'text/html');
    expect(mailObject.content, envelope.html);
  }, skip: 'Mailer package sends invalid MIME');

  test('require mail from', () async {
    var socket = await Socket.connect(server.address, server.port);
    var s = socket.asBroadcastStream();
    stdout.addStream(s);
    var lines = new StreamQueue<String>(
        s.transform(utf8.decoder).transform(const LineSplitter()));

    // Await 220
    await lines.next;

    socket.writeln('HELO dart_test');
    await lines.next; // Await 250

    // Send RCPT TO without MAIL FROM
    socket.writeln('RCPT TO:<foo@bar.com>');
    var line = await lines.next;
    expect(line, startsWith('503'));
    await socket.close();
  });

  test('require rcpt to', () async {
    var socket = await Socket.connect(server.address, server.port);
    var s = socket.asBroadcastStream();
    stdout.addStream(s);
    var lines = new StreamQueue<String>(
        s.transform(utf8.decoder).transform(const LineSplitter()));

    // Await 220
    await lines.next;

    socket.writeln('HELO dart_test');
    await lines.next; // Await 250

    socket.writeln('MAIL FROM:<foo@bar.com>');
    await lines.next; // Await 250

    // Send DATA without RCPT TO
    socket.writeln('DATA');
    var line = await lines.next;

    expect(line, startsWith('554'));
    await socket.close();
  });
}


lib/src/protocol.dart
part of smtp;

abstract class SmtpRequest implements IOSink {
  String get method;
  List<String> get arguments;

  Future close({int statusCode: 221, String reasonPhrase: 'Bye'});
}

abstract class SmtpMailObject {
  SmtpConnectionInfo get connectionInfo;

  bool get supportsSmtpExtensions;

  SmtpEnvelope get envelope;

  String get content;

  Stream<MimeMultipart> get mimeMultiparts;

  Future close({int statusCode: 221, String reasonPhrase: 'Bye'});
}

abstract class SmtpEnvelope {
  String get originatorAddress;

  List<String> get recipientAddresses;

  SmtpHeaders get headers;
}

abstract class SmtpHeaders {
  MediaType get contentType;

  DateTime get date;

  String get from;

  String get to;

  List<String> get cc;

  List<String> get bcc;

  String get subject;

  Map<String, String> toMap();
}

abstract class SmtpConnectionInfo {
  InternetAddress get remoteAddress;

  String get remoteHostname;

  int get localPort;

  int get remotePort;
}

class _SmtpRequestImpl implements SmtpRequest {
  final String method;
  final List<String> arguments;
  final Socket socket;
  final StreamSubscription sub;

  _SmtpRequestImpl(this.method, this.arguments, this.socket, this.sub);

  @override
  Future close({int statusCode: 221, String reasonPhrase: 'Bye'}) async {
    socket.writeln('$statusCode $reasonPhrase');
    sub.resume();
  }

  @override
  Encoding get encoding => socket.encoding;

  void set encoding(Encoding encoding) {
    socket.encoding = encoding;
  }

  @override
  Future get done => socket.done;

  @override
  Future flush() => socket.flush();

  @override
  Future addStream(Stream<List<int>> stream) => socket.addStream(stream);

  @override
  void addError(error, [StackTrace stackTrace]) =>
      socket.addError(error, stackTrace);

  @override
  void writeCharCode(int charCode) => socket.writeCharCode(charCode);

  @override
  void writeln([Object obj = ""]) => socket.writeln(obj);

  @override
  void writeAll(Iterable objects, [String separator = ""]) =>
      socket.writeAll(objects, separator);

  @override
  void write(Object obj) => socket.write(obj);

  @override
  void add(List<int> data) => socket.add(data);
}

class _SmtpMailObjectImpl implements SmtpMailObject {
  final SmtpConnectionInfo connectionInfo;
  final bool supportsSmtpExtensions;
  final SmtpEnvelope envelope;
  final String content;
  final Socket socket;

  _SmtpMailObjectImpl(this.connectionInfo, this.supportsSmtpExtensions,
      this.envelope, this.content, this.socket);

  @override
  Future close({int statusCode: 221, String reasonPhrase: 'Bye'}) {
    // Send Bye
    socket.writeln('$statusCode $reasonPhrase');
    return socket.close();
  }

  @override
  Stream<MimeMultipart> get mimeMultiparts {
    if (envelope.headers.contentType?.type != 'multipart') {
      throw new FormatException(
          'Can only parse a multipart/* body into separate parts.');
    }

    var boundary = envelope.headers.contentType.parameters['boundary'];

    if (boundary?.trim()?.isNotEmpty != true) {
      throw new FormatException(
          'Cannot parse body into parts; no boundary parameter was specified in the `Content-Type` header.');
    }

    var stream = new Stream<List<int>>.fromIterable([utf8.encode(content)]);
    return stream.transform(new MimeMultipartTransformer(boundary));
  }
}

class _SmtpEnvelopeImpl implements SmtpEnvelope {
  final String originatorAddress;
  final List<String> recipientAddresses;
  final SmtpHeaders headers;

  _SmtpEnvelopeImpl(
      this.originatorAddress, this.recipientAddresses, this.headers);
}

class _SmtpHeadersImpl implements SmtpHeaders {
  final Map<String, String> headers;

  List<String> _cc, _bcc;
  MediaType _contentType;
  DateTime _date;

  _SmtpHeadersImpl(Map<String, String> headers)
      : this.headers = new CaseInsensitiveMap.from(headers);

  MediaType get contentType =>
      _contentType ??= (headers.containsKey('Content-Type')
          ? new MediaType.parse(headers['Content-Type'])
          : null);

  @override
  DateTime get date {
    return _date ??=
        (headers.containsKey('Date') ? HttpDate.parse(headers['Date']) : null);
  }

  @override
  String get from => headers['From'];

  @override
  String get to => headers['To'];

  @override
  String get subject => headers['subject'];

  @override
  List<String> get cc =>
      _cc ??= (headers.containsKey('Cc') ? headers['Cc'].split(',') : []);

  @override
  List<String> get bcc =>
      _bcc ??= (headers.containsKey('Bcc') ? headers['Bcc'].split(',') : []);

  @override
  Map<String, String> toMap() => headers;
}

class _SmtpConnectionInfoImpl implements SmtpConnectionInfo {
  final InternetAddress remoteAddress;
  final String remoteHostname;
  final int localPort, remotePort;

  _SmtpConnectionInfoImpl(
      this.remoteAddress, this.remoteHostname, this.localPort, this.remotePort);
}


lib/src/server.dart
part of smtp;

abstract class SmtpServer extends Stream<SmtpRequest> {
  String hostname = Platform.localHostname;
  String greeting = '';

  static Future<SmtpServer> bind(address, int port,
      {int backlog: 0, bool v6Only: false, bool shared: false}) async {
    var socket = await ServerSocket.bind(address, port,
        backlog: backlog, v6Only: v6Only, shared: shared);
    return new _SmtpServerImpl(
        socket, socket.address, socket.port, socket.close);
  }

  static Future<SmtpServer> bindSecure(
      address, int port, SecurityContext context,
      {int backlog: 0,
      bool v6Only: false,
      bool shared: false,
      bool requestClientCertificate: false,
      bool requireClientCertificate: false,
      List<String> supportedProtocols}) async {
    var socket = await SecureServerSocket.bind(address, port, context,
        backlog: backlog,
        v6Only: v6Only,
        shared: shared,
        requestClientCertificate: requestClientCertificate,
        requireClientCertificate: requireClientCertificate,
        supportedProtocols: supportedProtocols);
    return new _SmtpServerImpl(
        socket, socket.address, socket.port, socket.close);
  }

  Stream<SmtpMailObject> get mailObjects;

  InternetAddress get address;

  int get port;

  Future close();
}

class _SmtpServerImpl extends SmtpServer {
  static final RegExp _header = new RegExp(r'([^:]+): ([^$]+)');
  static final RegExp _helo =
      new RegExp(r'(HELO|EHLO) ([^$]+)', caseSensitive: false);
  static final RegExp _mailFrom =
      new RegExp(r'MAIL FROM:<([^>]+)>', caseSensitive: false);
  static final RegExp _rcptTo =
      new RegExp(r'RCPT TO:<([^>]+)>', caseSensitive: false);
  final Stream<Socket> stream;
  final InternetAddress address;
  final int port;
  final Future Function() closeFunction;
  final StreamController<SmtpMailObject> _mailObjects = new StreamController();
  final StreamController<SmtpRequest> _stream = new StreamController();
  StreamSubscription _sub;

  _SmtpServerImpl(this.stream, this.address, this.port, this.closeFunction) {
    _sub = stream.listen(handleSocket,
        onError: _mailObjects.addError, onDone: _mailObjects.close);
  }

  @override
  Stream<SmtpMailObject> get mailObjects {
    return _mailObjects.stream;
  }

  Future close() async {
    _sub.cancel();
    _mailObjects.close();
    _stream.close();
    await closeFunction();
  }

  @override
  StreamSubscription<SmtpRequest> listen(void onData(SmtpRequest event),
      {Function onError, void onDone(), bool cancelOnError}) {
    return _stream.stream.listen(onData,
        onError: onError, onDone: onDone, cancelOnError: cancelOnError);
  }

  handleSocket(Socket socket) async {
    // Send 220
    socket.writeln('220 $hostname $greeting'.trim());

    var lineStream =
        socket.transform(utf8.decoder).transform(const LineSplitter());

    var interceptedLines = new StreamController<String>();
    bool shouldIntercept = true, closed = false;

    StreamSubscription sub;

    sub = lineStream.listen((line) {
      if (!shouldIntercept)
        interceptedLines.add(line);
      else if (line.isNotEmpty) {
        var split = line.split(' ');
        var method = split[0].toUpperCase(), arguments = split.skip(1).toList();

        if (method == 'HELO' ||
            method == 'EHLO' ||
            method == 'MAIL' ||
            method == 'RCPT' ||
            method == 'RSET' ||
            method == 'DATA')
          interceptedLines.add(line);
        else if (method == 'QUIT') {
          closed = true;
          sub.cancel();
          interceptedLines.close();
          socket.close();
        } else if (method == 'NOOP') {
          // Do nothing; No-op.
        } else {
          var request = new _SmtpRequestImpl(method, arguments, socket, sub);
          sub.pause();
          _stream.add(request);
        }
      }
    }, onError: interceptedLines.addError, onDone: interceptedLines.close);

    var lines = new StreamQueue<String>(interceptedLines.stream);

    // Wait for HELO
    SmtpConnectionInfo connectionInfo;
    bool supportsSmtpExtensions;

    while (await lines.hasNext) {
      var line = await lines.next;
      var heloMatch = _helo.firstMatch(line);

      if (heloMatch != null) {
        connectionInfo = new _SmtpConnectionInfoImpl(
            socket.remoteAddress, heloMatch[2], port, socket.remotePort);
        supportsSmtpExtensions = heloMatch[1].toLowerCase() == 'ehlo';
        break;
      }
    }

    if (closed) return;

    if (connectionInfo == null) {
      // If HELO is never sent, just close the connection.
      socket.close();
      return;
    }

    // Send a greeting
    socket.writeln('250 $hostname, I am glad to meet you');

    String mailFrom;
    List<String> rcptTo = [];

    while (await lines.hasNext) {
      var line = await lines.next;

      if (mailFrom == null) {
        var m = _mailFrom.firstMatch(line);

        if (m == null) {
          socket.writeln('503 command out of sequence');
          await socket.close();
          return;
        }

        mailFrom = m[1];
        socket.writeln('250 Ok');
      } else {
        var m = _rcptTo.firstMatch(line);
        if (m != null) {
          rcptTo.add(m[1]);
          socket.writeln('250 Ok');
        } else if (line == 'DATA') {
          if (rcptTo.isEmpty) {
            socket.writeln('554 no valid recipients');
            await socket.close();
            return;
          }

          shouldIntercept = false;
          socket.writeln('354 End data with <CR><LF>.<CR><LF>');
          break;
        } else {
          // If we receive unrecognized data, close the socket.
          socket.close();
        }
      }
    }

    // Read headers
    Map<String, String> headers = {};

    while (await lines.hasNext) {
      var line = await lines.next;

      if (line.isEmpty) break;

      var m = _header.firstMatch(line);

      if (m == null) {
        // If something other than a header is sent, just close the socket.
        socket.close();
        return;
      } else {
        headers[m[1]] = m[2];
      }
    }

    // Read message
    var message = new StringBuffer();

    while (await lines.hasNext) {
      var line = await lines.next;

      if (line == '.')
        break;
      else if (line.startsWith('.'))
        message.writeln(line.substring(1));
      else
        message.writeln(line);
    }

    // Create request
    var mailObject = new _SmtpMailObjectImpl(
      connectionInfo,
      supportsSmtpExtensions,
      new _SmtpEnvelopeImpl(
        mailFrom,
        rcptTo,
        new _SmtpHeadersImpl(headers),
      ),
      message.toString(),
      socket,
    );

    _mailObjects.add(mailObject);
  }
}
